src/independentlylinearizedutils.jl

```diff
@@ -2,30 +2,130 @@ using SciMLBase
 
 export IndependentlyLinearizedSolution
 
+
+"""
+    CachePool(T, alloc; thread_safe = true)
+
+Simple memory-reusing cache that allows us to grow a cache and keep
+re-using those pieces of memory (in our case, typically `u` vectors)
+until the solve is finished.  By default, this datastructure is made
+to be thread-safe by locking on every acquire and release, but it
+can be made thread-unsafe (and correspondingly faster) by passing
+`thread_safe = false` to the constructor.
+
+While manual usage with `acquire!()` and `release!()` is possible,
+most users will want to use `@with_cache`, which provides lexically-
+scoped `acquire!()` and `release!()` usage automatically.  Example:
+
+```julia
+us = CachePool(Vector{S}, () -> Vector{S}(undef, num_us); thread_safe=false)
+@with_cache us u_prev begin
+    @with_cache us u_next begin
+        # perform tasks with these two `u` vectors
+    end
+end
+```
+"""
+mutable struct CachePool{T, THREAD_SAFE}
+    pool::Vector{T}
+    alloc::Function
```

Comment:
```suggestion
    const pool::Vector{T}
    const alloc::Function
```

I don't have a terribly specific concern in mind here, but this should make it easier for the compiler to infer/inline the `alloc()`. For example, if the cachepool gets passed into any non-inlined function where we can't observe modifications to the `CachePool`.

Reply:
Will this work even though `alloc` is typed just as a `Function` and not as a parametric type of the precise function it is?  My understanding is that the way this is now, `getproperty(cache_pool, :alloc)` is inferred to return an abstract type (`Function`) so we're never going to be able to inline the result.

Reply:
You can't inline the result in any callee function, but you can at least inline it anywhere that you can see the construction of the `CachePool` or anywhere that receives the cachepool with the field as a known `Const` (i.e. with a `PartialStruct` return type for inference)

It essentially makes the field eligible for const-prop, rather than just SROA

Reply:
Ah, right, our `PartialStruct` stuff would get around this!  That's so neat!  Yeah, I doubt that's going to be the case in any of our workloads (since it is constructed at a very high level, then passed through many many layers of APIs) but there's no reason not to do this!